typing/typedecl.ml

```diff
@@ -1782,17 +1782,40 @@ let update_decl_jkind env dpath decl =
       type_jkind
   in
 
+  let allow_any_crossing =
+    Builtin_attributes.has_unsafe_allow_any_mode_crossing decl.type_attributes
+  in
+
   (* check that the jkind computed from the kind matches the jkind
      annotation, which was stored in decl.type_jkind *)
   if new_jkind != decl.type_jkind then
     (* CR layouts v2.8: Consider making a function that doesn't compute
        histories for this use-case, which doesn't need it. *)
-    begin match Jkind.sub_jkind_l new_jkind decl.type_jkind with
-    | Ok _ -> ()
+    begin match Jkind.sub_jkind_l ~allow_any_crossing new_jkind decl.type_jkind with
+    | Ok _ ->
+      if allow_any_crossing then
+        { new_decl with
+          (* If the user is asking us to allow any crossing, we use the modal bounds from
+             the annotation rather than the modal bounds inferred from the type_kind.
+             However, we /only/ take the modal bounds, not the layout - because we still
+             want to be able to eg locally use a type declared as layout [any] as [value]
+             if that's its actual layout! *)
+          type_jkind =
+            { new_decl.type_jkind with
+              jkind =
+                { new_decl.type_jkind.jkind with
```

Comment:
This is accessing private fields. The only reason this is possible is because of the sad jkind_types.ml dance. Maybe refactor to use a function to update the bounds?